README.md

```diff
@@ -0,0 +1,396 @@
+# What's This?
+
+This repo is an experiment in what a component-based Agent might look like.
+It's not intended to actually behave like an Agent -- all non-component implementation is fake.
+But it is intended to encompass the current design of the Agent, focusing on the different sorts of interconnections between components required.
+
+This provides a loosely-sketched vision of what we will want to get to with the Agent, encouraging discussion now, during the design phase, but with concrete code to look at.
+
+It is also a chance to try out some "interesting" ideas (such as config reducers) that would be much more difficult to sketch out in a real Agent.
+Maybe some of these are bad ideas, maybe they are things we can do later, or maybe we want to do them first thing.
+
+You can see the set of components in [`COMPONENTS.md`](./COMPONENTS.md).
+
+# Building
+
+To build, run `inv build`.
+This will build all binaries in the root of the repository.
+
+To regenerate files in the repo, run `inv generate`.
+
+# Component Guidelines
+
+What follows are draft guidelines for writing components.
+This repository follows these guidelines.
+
+If these guidelines are adopted, then the expectation is that all components would follow them.
+This reduces cognitive load when using components -- no need to remember one component's peculiarities.
+It also allows Agent-wide changes, where we make the same formulaic change to each component.
+If a situation arises that contradicts the guidelines, then we can update the guidelines (and change all affected components).
+
+## Prerequisites
+
+You should be familiar with [`fx`](https://pkg.go.dev/go.uber.org/fx), the tool used for dependency injection.
+
+## Component Definition
+
+A component is defined in a dedicated package under `comp/`, with the following defined in `component.go`:
+
+ * Extensive package-level documentation.
+   This should define, as precisely as possible, the behavior of the component, acting as a contract on which users of the component may depend.
```

Comment:
I agree, but Go isn't really designed that way.  Most of the Go stdlib depends on careful reading of its documentation.

Reply:
> I agree, but Go isn't really designed that way
Indeed

I like APIs with minimal documentation because the API is designed in a way to prevent bad usage (which is often challenging and sometime impossible).

Reply:
https://www.youtube.com/watch?v=JQYnFyG7A8c was a nice talk on the topic :)